src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/GCHandle.T.cs

```diff
@@ -0,0 +1,100 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Represents an strongly-typed opaque, GC handle to a managed object.
+    /// A GC handle is used when an object reference must be reachable from
+    /// unmanaged memory.
+    /// </summary>
+    /// <remarks>
+    /// <see cref="GCHandle{T}"/> corresponds to Normal roots.
+    /// For Weak and WeakTrackResurrection, see <see cref="WeakGCHandle{T}"/>.
+    /// For Pinned, see <see cref="PinnedGCHandle{T}"/>.
+    /// </remarks>
+    /// <seealso cref="GCHandle" />
+    /// <typeparam name="T">The type of the object this <see cref="GCHandle{T}"/> tracks to.</typeparam>
+    public struct GCHandle<T> : IDisposable
+        where T : class?
+    {
+        // The actual integer handle value that the EE uses internally.
+        private IntPtr _handle;
+
+        /// <summary>
+        /// Allocates a handle for the specified object.
+        /// </summary>
+        /// <param name="target">The object that uses the <see cref="GCHandle{T}"/>.</param>
+        public GCHandle(T target)
+        {
+            _handle = GCHandle.InternalAlloc(target, GCHandleType.Normal);
+        }
+
+        private GCHandle(IntPtr handle) => _handle = handle;
+
+        /// <summary>Determine whether this handle has been allocated or not.</summary>
+        public readonly bool IsAllocated => _handle != IntPtr.Zero;
+
+        /// <summary>Gets or sets the object this handle represents.</summary>
+        public readonly T Target
+        {
+            get
+            {
+                IntPtr handle = _handle;
+                GCHandle.ThrowIfInvalid(_handle);
```

Comment:
All these checks should be removed. Using any instance method on an instance that's not allocated is explicitly not supported. As long as we can throw an exception (eg. this should throw NRE) so the behavior is still consistent, that's sufficient.

Reply:
> As long as we can throw an exception (eg. this should throw NRE)

This is the questionable part - the current implementation uses FCall in some cases, and the ability to throw exception from FCall will soon be removed. In other words, no NRE can be thrown, at least for some members.

With that said, providing invalid value via `FromIntPtr` will still cause hard AV in unmanaged code.

Reply:
At least for the getter, on CoreCLR and NAOT this should just be `*(T*)_handle`, which should already throw NRE implicitly.

Reply:
Maybe an implicit read (`_ = *(byte*)value`) can be applied with lowest overhead? I'm not worried about garbage value from `FromIntPtr`, but make `new GCHandle<T>{ Target = obj }` crashing doesn't look good.

Reply:
What do you mean by implicit read? You need to read anyway, since you're returning a value. What I'm saying is the whole getter should literally just be `get => *(T*)_handle`. That will automatically NRE if `handle` is `null`, which is all we need.

Reply:
I mean adding a read in managed code in setter. Currently the entire setter is executed in unmanaged code and will cause AV, which won't be caught as NRE.

Reply:
Oh I see what you mean, I thought you were talking about the other accessor. That would seem fine to me, but I'll defer to Jan and others to confirm 🙂